programs/stake/src/stake_state.rs

```diff
@@ -756,12 +756,24 @@ impl Stake {
         stake_history: &StakeHistory,
         config: &Config,
     ) -> Result<(), StakeError> {
-        // can't redelegate if stake is active.  either the stake
-        //  is freshly activated or has fully de-activated.  redelegation
-        //  implies re-activation
+        // If stake is currently active:
         if self.stake(clock.epoch, Some(stake_history), true) != 0 {
-            return Err(StakeError::TooSoonToRedelegate);
+            // If pubkey of new voter is the same as current,
+            // and we have not deactivated any stake yet, we un-deactivate stake.
+            if self.delegation.voter_pubkey == *voter_pubkey
+                && clock.epoch == self.delegation.deactivation_epoch
+            {
```

Comment:
This will need to be activated by feature.  [This PR](https://github.com/solana-labs/solana/pull/15495/commits/f56ca2ee20d7592bbd4f6c50b3197e8cc5e1b84e) is a pretty good  example of the necessary plumbing

Reply:
What's the stake program work? is it anything v2-related?

Reply:
I have a small stake-program fix I'd like to piggy-back in on a feature as well.

Reply:
Please rebase on https://github.com/solana-labs/solana/pull/17353 and use this feature.
I hold the key for this one

Reply:
> I have a small stake-program fix I'd like to piggy-back in on a feature as well.

@CriesofCarrots - Do it!  The time is now :)

Reply:
Does holding the private key for a program grant you special powers?

Reply:
> Does holding the private key for a program id grant special powers?

I assume you mean feature id?  Yeah it means you can break cluster consensus if the feature activation is timed poorly :)

Reply:
So the creator of the feature has the full authority to activate the feature? Is a program's authority set to automatically be the program's pubkey?

Reply:
> So the creator of the feature has the full authority to activate the feature? 

For these kinds of technical features that don't require community governance, yes.

For features that do require governance, we have https://spl.solana.com/feature-proposal

There's a comment at the top of `sdk/program/src/feature.rs` that describes how this works a little